src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -906,5 +914,141 @@ private static void ReadToEndOfCurrentObject(JsonReader reader)
+        /// The purpose of this class is to avoid permanently storing duplicate strings which were read in from JSON.
+        /// </summary>
+        private class WeakReferenceTable
+        {
+            public static readonly WeakReferenceTable Instance = new WeakReferenceTable();
+
+            private WeakReferenceTable()
+            {
+            }
+
+            private readonly WeakValueDictionary<string, string> _table = new WeakValueDictionary<string, string>();
+
+            public string Intern(string str)
+            {
+                if (_table.TryGetValue(str, out var value))
+                {
+                    return value;
+                }
+                else
+                {
+                    _table.Add(str, str);
+                    return str;
+                }
+            }
+
+            // ConditionalWeakTable won't work for us because it only compares to Object.ReferenceEquals.
+            private class WeakValueDictionary<TKey, TValue> where TValue : class
+            {
+                // Your basic project.razor.json using our template have ~800 entries,
+                // so lets start the dictionary out large
+                private int _capacity = 400;
+
+                // This might not acctually need to be concurrent?
+                private readonly ConcurrentDictionary<TKey, WeakReference> _dict = new ConcurrentDictionary<TKey, WeakReference>();
+
+                public bool TryGetValue(TKey key, out TValue value)
+                {
+                    if (key is null)
+                    {
+                        throw new ArgumentNullException(nameof(key));
+                    }
+
+                    if (_dict.TryGetValue(key, out var weak))
+                    {
+                        value = weak.Target as TValue;
+
+                        if (value is null)
+                        {
+                            // This means we lost all references to the value, time to remove the entry.
+                            _dict.TryRemove(key, out _);
+                            return false;
+                        }
+
+                        return true;
+                    }
+
+                    // We didn't find the value
+                    value = null;
+                    return false;
+                }
+
+                public void Add(TKey key, TValue value)
+                {
+                    if (key is null)
+                    {
+                        throw new ArgumentNullException(nameof(key));
+                    }
+
+                    if (value is null)
+                    {
+                        throw new ArgumentNullException(nameof(value));
+                    }
+
+                    if (_dict.Count == _capacity)
+                    {
+                        Cleanup();
+
+                        if (_dict.Count == _capacity)
+                        {
+                            _capacity = _dict.Count * 2;
```

Comment:
You can unintentionally run this more then once if you have multiple threads trying to add simultaneously and all running through cleanup. Aka, imagine multiple threads at line 1013 which is possible given a full dictionary of all still "live" strings" you'd result in increasing the capacity twice.

I think ultimately this is OK; however, I'd make the backing dictionary non-concurrent and enforce concurrency at the consumer layer.